### PR TITLE
Drop unnecessary tokens in GPT2Model generation

### LINKED ISSUE #38977: LMHead is processing redundant tokens in prefill
While using `GPT2LMHeadModel.generate()` and compare its performance with vLLM, I noticed a significant inefficiency in the `forward()` implementation of many huggingface models. For example, in the `GPT2LMHeadModel.forward`, `self.lm_head` is applied to all token hidden states, even when called from the `generate()` method, where only the logits of the last token are needed for next-token prediction. This computes logits over the entire sequence and can introduce significant overhead.

```py
# src/transformers/models/gpt2/modeling_gpt2.py, line 1233
lm_logits = self.lm_head(hidden_states)
```

Suggested Fix: add a conditional branch in forward() to slice the hidden states before computing logits if it’s a generation step.

### PR DESCRIPTION
**Pull request by null-pointer-access**
# What does this PR do?

In the current GPT2 implementation, the LMHead module processes all tokens during prefill, even though only the final token’s output is used for generation. This PR aligns the behavior with LlamaModel by computing the LMHead output only for the last token, reducing unnecessary computation during prefill.

Fixes #38977 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case. #38977 
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Review by zucchini-nlp — APPROVED** [review]
Perfect, looks good to me! 

Would be nice to do a pass over old common models and update them as well, we will leave it to later PRs :)
